File: app/lib/models/question.dart
import 'choice_option.dart';

enum QuestionType {
  choice,
  shortAnswer;

  factory QuestionType.fromJson(String value) {
    return switch (value) {
      'choice' => QuestionType.choice,
      'short_answer' || 'shortAnswer' => QuestionType.shortAnswer,
      _ => QuestionType.shortAnswer,
    };
  }

  String toJson() => switch (this) {
    QuestionType.choice => 'choice',
    QuestionType.shortAnswer => 'short_answer',
  };
}

class Question {
  final int id;
  final String title;
  final String description;
  final QuestionType type;
  final String category;
  final int difficulty;
  final List<ChoiceOption>? options;

  const Question({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.category,
    required this.difficulty,
    this.options,
  });

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      id: json['id'] as int,
      title: json['title'] as String,
      description: json['description'] as String,
      type: QuestionType.fromJson(json['type'] as String),
      category: json['category'] as String,
      difficulty: json['difficulty'] as int,
      options: (json['options'] as List<dynamic>?)
          ?.map((e) => ChoiceOption.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: backend/app/models/question.py
import enum
import uuid
from datetime import datetime, timezone
from typing import List, Optional

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class QuestionType(str, enum.Enum):
    choice = "choice"
    short_answer = "short_answer"


class Question(Base):
    __tablename__ = "questions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(String(500))
    title_en: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    description: Mapped[str] = mapped_column(Text)
    description_en: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    type: Mapped[str] = mapped_column(String(20))
    category: Mapped[str] = mapped_column(String(100), index=True)
    difficulty: Mapped[int] = mapped_column(Integer)
    is_free: Mapped[bool] = mapped_column(Boolean, default=True)
    metadata_: Mapped[Optional[dict]] = mapped_column("metadata", JSON, nullable=True)

    options: Mapped[List["ChoiceOption"]] = relationship(back_populates="question", lazy="selectin")


class ChoiceOption(Base):
    __tablename__ = "choice_options"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    question_id: Mapped[int] = mapped_column(Integer, ForeignKey("questions.id"))
    content: Mapped[str] = mapped_column(String(1000))
    content_en: Mapped[Optional[str]] = mapped_column(String(1000), nullable=True)
    is_interesting: Mapped[bool] = mapped_column(Boolean, default=False)
    ai_comment: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    ai_comment_en: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    question: Mapped["Question"] = relationship(back_populates="options")


class UserQuestion(Base):
    __tablename__ = "user_questions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[str] = mapped_column(String(36), ForeignKey("users.id"))
    title: Mapped[str] = mapped_column(String(500))
    description: Mapped[str] = mapped_column(Text)
    category: Mapped[str] = mapped_column(String(100))
    status: Mapped[str] = mapped_column(String(20), default="pending")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    reviewed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    user = relationship("User", lazy="selectin")
